File: to-fhir-py/src/to_fhir/id_utils.py
"""Utilities for deterministic FHIR resource ids."""

from __future__ import annotations

import hashlib

from fhir.resources.R4B.identifier import Identifier

#: The hash algorithm used unless a different :mod:`hashlib` algorithm is requested.
DEFAULT_ALGORITHM = "sha256"


def from_identifier(
    identifier: Identifier,
    resource_type: str | None = None,
    *,
    algorithm: str = DEFAULT_ALGORITHM,
) -> str:
    """Compute a deterministic id from a FHIR identifier by hashing its system and value.

    Args:
        identifier: The FHIR identifier to compute the id from.
        resource_type: When given, the id is returned as the relative reference
            ``<resource_type>/<id>`` instead of the bare id.
        algorithm: The :mod:`hashlib` algorithm to hash with, e.g. ``"sha256"``.

    Returns:
        The deterministic id, optionally qualified with ``resource_type``.

    Raises:
        ValueError: The identifier's system or value is missing or blank.
    """
    if not (identifier.system or "").strip():
        raise ValueError("Identifier system must not be blank.")

    if not (identifier.value or "").strip():
        raise ValueError(f"Identifier value must not be blank. System: {identifier.system}")

    digest = hashlib.new(algorithm, f"{identifier.system}|{identifier.value}".encode())
    resource_id = digest.hexdigest()

    if resource_type is None:
        return resource_id

    return f"{resource_type}/{resource_id}"

File: to-fhir-cs/ToFhir/IdUtils.cs
using System.Security.Cryptography;
using System.Text;
using Hl7.Fhir.Model;
using Hl7.Fhir.Utility;

namespace ToFhir;

/// <summary>Utility class for FHIR resource ids.</summary>
public static class IdUtils
{
    /// <summary>
    /// Computes a deterministic id from a FHIR <see cref="Identifier"/> by hashing its system and
    /// value with SHA-256.
    /// </summary>
    /// <param name="identifier">The FHIR identifier to compute the id from.</param>
    /// <returns>A deterministic id derived from the identifier's system and value.</returns>
    /// <exception cref="ArgumentException">
    /// The identifier's system or value is <c>null</c>, empty or whitespace.
    /// </exception>
    public static string FromIdentifier(Identifier identifier) =>
        FromIdentifier(identifier, HashAlgorithmName.SHA256);

    /// <summary>
    /// Computes a deterministic id from a FHIR <see cref="Identifier"/> using the given hash
    /// algorithm.
    /// </summary>
    /// <param name="identifier">The FHIR identifier to compute the id from.</param>
    /// <param name="hashAlgorithm">The hash algorithm to use, e.g. <see cref="HashAlgorithmName.SHA256"/>.</param>
    /// <returns>A deterministic id derived from the identifier's system and value.</returns>
    /// <exception cref="ArgumentException">
    /// The identifier's system or value is <c>null</c>, empty or whitespace.
    /// </exception>
    public static string FromIdentifier(Identifier identifier, HashAlgorithmName hashAlgorithm)
    {
        ArgumentNullException.ThrowIfNull(identifier);

        if (string.IsNullOrWhiteSpace(identifier.System))
        {
            throw new ArgumentException("Identifier system must not be blank.", nameof(identifier));
        }

        if (string.IsNullOrWhiteSpace(identifier.Value))
        {
            throw new ArgumentException(
                $"Identifier value must not be blank. System: {identifier.System}",
                nameof(identifier)
            );
        }

        var hash = CryptographicOperations.HashData(
            hashAlgorithm,
            Encoding.UTF8.GetBytes($"{identifier.System}|{identifier.Value}")
        );

        return Convert.ToHexStringLower(hash);
    }

    /// <summary>
    /// Computes a deterministic, type-qualified id from a FHIR <see cref="Identifier"/> by hashing
    /// its system and value with SHA-256.
    /// </summary>
    /// <param name="identifier">The FHIR identifier to compute the id from.</param>
    /// <param name="resourceType">The resource type to qualify the id with.</param>
    /// <returns>A relative reference of the form <c>ResourceType/id</c>.</returns>
    /// <exception cref="ArgumentException">
    /// The identifier's system or value is <c>null</c>, empty or whitespace.
    /// </exception>
    public static string FromIdentifier(Identifier identifier, ResourceType resourceType) =>
        FromIdentifier(identifier, resourceType, HashAlgorithmName.SHA256);

    /// <summary>
    /// Computes a deterministic, type-qualified id from a FHIR <see cref="Identifier"/> using the
    /// given hash algorithm.
    /// </summary>
    /// <param name="identifier">The FHIR identifier to compute the id from.</param>
    /// <param name="resourceType">The resource type to qualify the id with.</param>
    /// <param name="hashAlgorithm">The hash algorithm to use, e.g. <see cref="HashAlgorithmName.SHA256"/>.</param>
    /// <returns>A relative reference of the form <c>ResourceType/id</c>.</returns>
    /// <exception cref="ArgumentException">
    /// The identifier's system or value is <c>null</c>, empty or whitespace.
    /// </exception>
    public static string FromIdentifier(
        Identifier identifier,
        ResourceType resourceType,
        HashAlgorithmName hashAlgorithm
    ) => $"{resourceType.GetLiteral()}/{FromIdentifier(identifier, hashAlgorithm)}";
}
